gateway/src/apicast/policy/soap/soap.lua

```diff
@@ -0,0 +1,127 @@
+--- SOAP Policy
+-- This policy adds support for a very small subset of SOAP.
+-- This policy basically expects a SOAPAction URI in the SOAPAction header or
+-- the content-type header.
+-- The SOAPAction header is used in v1.1 of the SOAP standard:
+-- https://www.w3.org/TR/2000/NOTE-SOAP-20000508/#_Toc478383528, whereas the
+-- Content-Type header is used in v1.2 of the SOAP standard:
+-- https://www.w3.org/TR/soap12-part2/#ActionFeature
+-- The SOAPAction URI is matched against the mapping rules defined in the
+-- policy and calculates a usage based on that so it can be authorized and
+-- reported against 3scale's backend.
+
+local sub = string.sub
+local len = string.len
+local ipairs = ipairs
+local insert = table.insert
+local re = require('ngx.re')
+
+local MappingRule = require('apicast.mapping_rule')
+local Usage = require('apicast.usage')
+local mapping_rules_matcher = require('apicast.mapping_rules_matcher')
+
+local policy = require('apicast.policy')
+
+local _M = policy.new('SOAP policy')
+
+local soap_action_header = 'SOAPAction'
+local soap_action_ctype = 'application/soap+xml;'
+
+local new = _M.new
+
+local function starts_with(str, start)
+  return sub(str, 1, len(start)) == start
+end
+
+-- Extracts a SOAP action from the SOAPAction header. Returns nil when not
+-- present.
+local function soap_action_in_header(headers)
+  return headers[soap_action_header]
+end
+
+-- Extracts the SOAP action from a string that contains the parameters of a
+-- Content-Type header. The string has this format:
+-- a_param=x;action=soap_action;another_param=y
+-- This method returns the value of 'action' or nil when it's not present.
+local function soap_action_from_ctype_params(params)
+  local params_split = re.split(params, ";")
+
+  for _, param in ipairs(params_split) do
+    local header_param_split = re.split(param, "=")
+    local name = header_param_split[1]
+    local value = header_param_split[2]
+    if name == "action" then
+      return value
+    end
+  end
+
+  return nil
+end
+
+-- Extracts a SOAP action from the Content-Type header. In SOAP, the
+-- type/subtype is application/soap+xml, and the action is specified as a
+-- param in that header. When there is no SOAP action, this method returns nil.
+local function soap_action_in_ctype(headers)
+  local ctype = headers['Content-Type']
+
+  if ctype and starts_with(ctype, soap_action_ctype) then
```

Comment:
I think this won't be that easy 🤔 
The [RFC](https://tools.ietf.org/html/rfc7231#section-3.1.1.1) says those are all equivalent:

>   text/html;charset=utf-8
     text/html;charset=UTF-8
     Text/HTML;Charset="utf-8"
     text/html; charset="utf-8"

And looks like it can have whitespace before `;` too. Look at the definition: ` media-type = type "/" subtype *( OWS ";" OWS parameter )`. And `OWS` is optional white space.

Reply:
You're right :+1:

Reply:
Check last commit.